### ISSUE TITLE
etcd unavailable during startup disables admin apis, but etcd unavailable later does not

### ISSUE THREAD

**Issue by markmarijnissen** [minio/minio#9788]
## Current Behavior

**Use case 1**
Run minio in gateway mode with an unavailable etcd endpoint.

Minio shows an error, but keeps running:
```
Error: Unable to initialize etcd config: dial tcp: lookup minio-etcd.minio.svc.cluster.local on 10.64.0.10:53: no such host
       2: github.com/minio/minio@/cmd/config-current.go:334:cmd.lookupConfigs()
       1: github.com/minio/minio@/cmd/gateway-main.go:165:cmd.StartGateway()
```

Subsequent commands that rely on etcd fail:

- `mc admin policy list xxx` returns `mc: <ERROR> Cannot list policy. This 'admin' API is not supported by server in 'mode-gateway-gcs'.`

**Use Case 2**
Run minio in gateway mode with an available etcd endpoint. After starting minio successfully, shutdown the etcd server.

- `mc admin policy list xxx` still returns the policies (cached?)
- `mc admin policy add xxx yyy zzz.json` does return after a timeout with a descriptive error mesage. Logs also show that etcd is unreachable.

```
mc: <ERROR> Unable to add new policy. We encountered an internal error, please try again. (etcd is unreachable, please check your endpoints [http://minio-etcd.minio.svc.cluster.local:2379]).
```

```
Error: etcd is unreachable, please check your endpoints [http://minio-etcd.minio.svc.cluster.local:2379]
       5: github.com/minio/minio@/cmd/api-errors.go:1875:cmd.toAPIErrorCode()
       4: github.com/minio/minio@/cmd/admin-handlers.go:872:cmd.toAdminAPIErrCode()
       3: github.com/minio/minio@/cmd/admin-handlers.go:909:cmd.toAdminAPIErr()
       2: github.com/minio/minio@/cmd/admin-handlers-users.go:864:cmd.adminAPIHandlers.AddCannedPolicy()
       1: net/http/server.go:2007:http.HandlerFunc.ServeHTTP()
```

## Expected Behavior

**1) etcd configuration is not optional.** 
If minio fails to initialize etcd, it should either (a) keep attempting to initialize or (b) exit with a descriptive error message.

**2) Minio behavior should not depend on startup order**
If etcd is unavailable on startup, it disables all admin commands.
If etcd is unavailable later, it enables admin commands, and responds and logs with a descriptive error message. 

## Possible Solutions

a) Add a flag `MINIO_ETCD_REQUIRED=1` that exits minio when etcd is not available on startup, instead of disabling certain admin api's.

b) Keep attempting to initialize etcd.

## Steps to Reproduce (for bugs)

See "Current Behavior" above.

## Context
Minio GCS gateway with ectd running on kubernetes. Allows Data Scientist to build pipelines and access their data and store intermediate results and anaysis. As the credentials suddenly became unvalid, the microservice that uploads data did not work. This broke the realtime analysis pipeline that delivers results back to our end-users (on a smartphone app).

Took a while to trace back the error from "invalid credentials" to failing etcd initialization, as the error message was not helpful and the events were a few hours distant in time.

## Regression
No

## Your Environment
* Version used: **minio/minio:RELEASE.2020-06-03T22-13-49Z** (etcd: 3.3.10-amd64)
* Environment name and version:
* Server type and version: kubernetes
* Operating System and version: docker image
* Link to your project: n/a (private)

**Comment by markmarijnissen** [comment]
Related PR: https://github.com/minio/minio/pull/7097

**Comment by nitisht** [comment]
> Version used: minio/minio:RELEASE.2019-01-23T23-18-58Z (etcd: 3.3.10-amd64)

This is way old. Can you try with latest release?

**Comment by markmarijnissen** [comment]
Will do. What is the best way to check changelog/commits for any changes on this?

**Comment by markmarijnissen** [comment]
@nitisht upgraded to latest minio, behavior is similar.

**Comment by nitisht** [comment]
Thanks @markmarijnissen we'll take a look

**Comment by ktarplee** [comment]
I can confirm that I just saw this with RELEASE.2020-06-03T22-13-49Z.  I have seen both case 1 and case 2.  This is pretty important to fix for this mode (nas gateway with etcd) to work properly.  The failure mode is kind of a "brown out" problem where you loose some functionality but not all.

**Comment by harshavardhana** [comment]
There is a chicken and egg problem for us, if etcd is not available when the server startups we cannot really enable these handlers because we need an instance of etcd.Client initialized. 

etcd.Client library expects the server to be up when we do etcd.Client.New() - handling this would horrible hacks. It is worthwhile to wait for etcd to startup before NAS gateway and you should order them appropriately.

I don't think it's feasible to add this complexity to our code.

**Comment by ktarplee** [comment]
All we are asking is that if the etcd connection fails (during startup or after startup), then minio immediately exits with non-zero exit code.  Minio will be restarted by the supervisor (in k8s this is the replicaset or statefulset).  I'm not seeing why that is difficult.  A re-try would be a pain, but that is not necessary.

**Comment by harshavardhana** [comment]
> All we are asking is that if the etcd connection fails (during startup or after startup), then minio immediately exits with non-zero exit code. Minio will be restarted by the supervisor (in k8s this is the replicaset or statefulset). I'm not seeing why that is difficult. A re-try would be a pain, but that is not necessary.

@ktarplee not all deployments use k8s, not all our deployments are container. the design to bring the server up was for availability point of view i.e server is always up.

MinIO cannot expect every user to have k8s or containers deployed. We will explore and see other options are possible..

**Comment by ktarplee** [comment]
Any decent service supervisor does this for you.  For example supervisord, initd, and systemd will all try to restart failed services and they use raw processes (no containers).  This does not depend on k8s or containers.  

Is there some way to detect this "brown out" failure so we can manually restart minio?  

Can we have an environment variable MINIO_ETCD_REQUIRED that when true will kill minio if the etcd connection is broken?

**Comment by ktarplee** [comment]
I have a question.  When ETCD is not available, what else will fail within Minio?  Does Minio use ETCD for any synchronization that might cause it to not be able to serve data from specific buckets?  I think the secret keys and access keys for users are stored in etcd, right?  Are they cached by Minio so Minio is still functional without etcd but you can't add or remove users?

**Comment by harshavardhana** [comment]
etcd is only meant to store credentials and policies etcd is not used for anything else.

**Comment by ktarplee** [comment]
> etcd is only meant to store credentials and policies etcd is not used for anything else.

Thanks for the response @harshavardhana  

So when etcd goes down and Minio stays up, can buckets still be accessed that require credentials?  Does Minio keep a copy of the credentials and policies or does it hit etcd upon every request?  I'm trying to understand how crippled Minio is after it looses a connection to etcd. 

Obviously if ETCD is not up before Minio comes up and credentials are only stored in etcd, then nothing requiring credentials can be downloaded or uploaded.  So in this case, while the Minio server is up, no one can access their data.  Is that correct?

**Comment by harshavardhana** [comment]
> So when etcd goes down and Minio stays up, can buckets still be accessed that require credentials? Does Minio keep a copy of the credentials and policies or does it hit etcd upon every request? I'm trying to understand how crippled Minio is after it looses a connection to etcd.

yes

> Obviously if ETCD is not up before Minio comes up and credentials are only stored in etcd, then nothing requiring credentials can be downloaded or uploaded. So in this case, while the Minio server is up, no one can access their data. Is that correct?

yes

**Comment by markmarijnissen** [comment]
I proposed two solutions for this:

1. retry initialization until it succeeds
2. add a `MINIO_ETCD_REQUIRED` flag

I understand that 1 is not possible due to the fact you need a etcd client instance. I also understand your concerns in making the server crash default behavior. Therefore, the second solution seems easy and viable to implement. 

I find operating kubernetes much easier if I avoid ordering between workloads, and I'd rather rely on minio restarting until etcd is available.

If we can agree on solution #2 I can try to make a PR (I have no experience with go, compling it or writing tests, so it might be easier if somebody else is able to do this)

**Comment by harshavardhana** [comment]
Fixed by #9842
